### ISSUE TITLE
Test failures - "test timed out after 1h30m0s"

### ISSUE THREAD

**Issue by dnephin** [docker/docker#25028]
A bunch of tests runs are failing with `test timed out after 1h30m0s`. The timeout happens in different tests, but one common element is that it seems to be failing during a call to [os/exec](https://golang.org/pkg/os/exec/) which we often wrap with `github.com/docker/docker/pkg/integration.RunCommand()`

**Comment by AkihiroSuda** [comment]
`pkg/integration` is known to be racy https://github.com/docker/docker/issues/22965

**Comment by dnephin** [comment]
```
19:07:20 os.(*Process).wait(0xc822856600, 0xe, 0x0, 0x0)
19:07:20 	/usr/local/go/src/os/exec_unix.go:22 +0x105
19:07:20 os.(*Process).Wait(0xc822856600, 0x0, 0x0, 0x0)
19:07:20 	/usr/local/go/src/os/doc.go:49 +0x2d
19:07:20 os/exec.(*Cmd).Wait(0xc821ea88c0, 0x0, 0x0)
19:07:20 	/usr/local/go/src/os/exec/exec.go:396 +0x211
19:07:20 os/exec.(*Cmd).Run(0xc821ea88c0, 0x0, 0x0)
19:07:20 	/usr/local/go/src/os/exec/exec.go:262 +0x64
19:07:20 os/exec.(*Cmd).CombinedOutput(0xc821ea88c0, 0x0, 0x0, 0x0, 0x0, 0x0)
19:07:20 	/usr/local/go/src/os/exec/exec.go:453 +0x310
19:07:20 github.com/docker/docker/pkg/integration.RunCommandWithOutput(0xc821ea88c0, 0x0, 0x0, 0xc8201af180, 0x0, 0x0)
19:07:20 	/go/src/github.com/docker/docker/pkg/integration/utils.go:66 +0x41
19:07:20 github.com/docker/docker/integration-cli.runCommandWithOutput(0xc821ea88c0, 0x0, 0x0, 0x3, 0x0, 0x0)
19:07:20 	github.com/docker/docker/integration-cli/_test/_obj_test/utils.go:39 +0x47
19:07:20 github.com/docker/docker/integration-cli.getAllContainers(0x0, 0x0, 0x0, 0x0)
19:07:20 	github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:309 +0xf5
19:07:20 github.com/docker/docker/integration-cli.deleteAllContainers(0x0, 0x0)
19:07:20 	github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:321 +0x40
19:07:20 github.com/docker/docker/integration-cli.(*DockerSuite).TearDownTest(0x203f540, 0xc8207d11d0)
19:07:20 	/go/src/github.com/docker/docker/integration-cli/check_test.go:37 +0x1d
```

**Comment by dnephin** [comment]
```
net/http.(*body).Read(0xc8204f63c0, 0xc8202472c1, 0x53f, 0x53f, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/http/transfer.go:643 +0x110
github.com/docker/docker/pkg/ioutils.(*readCloserWrapper).Read(0xc820284600, 0xc8202472c1, 0x53f, 0x53f, 0x16, 0x0, 0x0)
        <autogenerated>:1 +0x82
bytes.(*Buffer).ReadFrom(0xc82057f4e0, 0x7f02d0e4de00, 0xc820284600, 0xc1, 0x0, 0x0)
        /usr/local/go/src/bytes/buffer.go:176 +0x23f
io/ioutil.readAll(0x7f02d0e4de00, 0xc820284600, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/io/ioutil/ioutil.go:33 +0x156
io/ioutil.ReadAll(0x7f02d0e4de00, 0xc820284600, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/io/ioutil/ioutil.go:42 +0x51
github.com/docker/docker/integration-cli.readBody(0x7f02d0e4ddd0, 0xc820284600, 0x0, 0x0, 0x0, 0x0, 0x0)
        github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:288 +0xbf
github.com/docker/docker/integration-cli.(*DockerSuite).TestBuildApiRemoteTarballContextWithCustomDockerfile(0x1fdf520, 0xc8202085a0)
        /go/src/github.com/docker/docker/integration-cli/docker_api_build_test.go:123 +0xa5c
```

**Comment by dnephin** [comment]
```
os/exec.(*Cmd).Run(0xc823359040, 0x0, 0x0)
        /usr/local/go/src/os/exec/exec.go:262 +0x64
github.com/docker/docker/pkg/integration.RunCommand(0xc823359040, 0xc8230f4140, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/integration/utils.go:161 +0x2f
github.com/docker/docker/integration-cli.runCommand(0xc823359040, 0x60, 0x0, 0x0)
        github.com/docker/docker/integration-cli/_test/_obj_test/utils.go:59 +0x39
github.com/docker/docker/integration-cli.deleteContainer(0xc820402d00, 0xc, 0x0, 0x0)
        github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:295 +0x246
github.com/docker/docker/integration-cli.deleteAllContainers(0x0, 0x0)
        github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:331 +0x1a9
github.com/docker/docker/integration-cli.(*DockerSuite).TearDownTest(0x1fe1d60, 0xc8232c4870)
```

**Comment by aaronlehmann** [comment]
Ran into this problem here: https://jenkins.dockerproject.org/job/Docker-PRs-experimental/21727/console

**Comment by dnephin** [comment]
Daemon from that run during the last test:
```
Calling GET /v1.25/exec/d8ab55f66a3df329e39d95ff1ce85925eac65122d2bd303952d9eb3f96890816/json"
Calling GET /v1.25/containers/json?all=1&filters=%7B%22status%22%3A%7B%22paused%22%3Atrue%7D%7D"
Calling GET /v1.25/containers/json?all=1"
Calling DELETE /v1.25/containers/b9b38d6c4275?force=1&v=1"
Sending 9 to b9b38d6c4275ec72de9b215fb58922201926825d35e82373c6c1b3322fa1f76f"
containerd: process exited" id=b9b38d6c4275ec72de9b215fb58922201926825d35e82373c6c1b3322fa1f76f pid=init status=137 systemPid=7850
libcontainerd: received containerd event: &types.Event{Type:\"exit\", Id:\"b9b38d6c4275ec72de9b215fb58922201926825d35e82373c6c1b3322fa1f76f\", Status:0x89, Pid:\"init\", Timestamp:(*timestamp.Timestamp)(0xc821640250)}"
Revoking external connectivity on endpoint two (085305fbabf96d901d23441fc66e876037144c9a1af54d05a25eb4766ad871b5)"
Releasing addresses for endpoint two's interface on network bridge"
ReleaseAddress(LocalDefault/172.18.0.0/16, 172.18.0.3)"
Calling DELETE /v1.25/containers/d0825ba8761c?force=1&v=1"
Sending 9 to d0825ba8761c4d5a7a5b090515a4a901ac937c4f9ab3d059966de27818cca243"
containerd: process exited" id=d0825ba8761c4d5a7a5b090515a4a901ac937c4f9ab3d059966de27818cca243 pid=init status=137 systemPid=7917
libcontainerd: received containerd event: &types.Event{Type:\"exit\", Id:\"d0825ba8761c4d5a7a5b090515a4a901ac937c4f9ab3d059966de27818cca243\", Status:0x89, Pid:\"init\", Timestamp:(*timestamp.Timestamp)(0xc8248dcf30)}"
Revoking external connectivity on endpoint one (e90bef4fdc4374a8acdb24d54f1ec9fcbd6ed9ac80a7317fdc33da7989f6e8c2)"
```

And after the timeout and force shutdown:
```
clean 69 unused exec commands"
Processing signal 'terminated'"
starting clean shutdown of all containers..."
Force shutdown daemon"                                                                                                                                                                                                                                                            libcontainerd: containerd connection state change: SHUTDOWN"
stopping containerd after receiving terminated"
```

I don't know if that "unused exec commands" is normal, but otherwise it seems like the daemon is just fine and responding to requests.

**Comment by dnephin** [comment]
Chatting with @cpuguy83  and looking at a few more of these, the request seems to be getting stuck between the "Revoking external connectivity" and "Releasing address", so likely in libnetwork.

**Comment by coolljt0725** [comment]
@dnephin  if it was stuck between `Revoking external connectivity` and `Releasing address`, pay attention to these two places https://github.com/docker/docker/blob/master/vendor/src/github.com/docker/libnetwork/drivers/bridge/bridge_store.go#L119 and https://github.com/docker/docker/blob/master/vendor/src/github.com/docker/libnetwork/endpoint_cnt.go#L141. There is a loop here, I once encountered  a similar issue  on my other project which use `libkv` and `libnetwork/datastore`, and the operation are similar with these two places, we hit stuck occasionally because of infinite retrying on updating store.

**Comment by cpuguy83** [comment]
We resolved this in 1.12.0 in #25191
